jlm/rvsdg/traverser.hpp

```diff
@@ -24,208 +24,201 @@ class Region;
 namespace detail
 {
 
-template<typename T>
+enum class TraversalNodestate
+{
+  ahead = -1,
+  frontier = 0,
+  behind = +1
+};
+
+/* support class to track traversal states of nodes */
+template<typename NodeType>
+class TraversalTracker final
+{
+public:
+  /** \brief Determines whether node has been visited already. */
+  bool
+  isNodeVisited(NodeType * node) const;
+
+  /** \brief Checks activation count whether node is ready for visiting. */
+  void
+  checkNodeActivation(NodeType * node, std::size_t threshold);
+
+  /** \brief Checks activation count whether node is no longer ready for visiting. */
+  void
+  checkNodeDeactivation(NodeType * node, std::size_t threshold);
+
+  /** \brief Marks a node visited if it is currently ready for visiting. */
+  void
+  checkMarkNodeVisitedIfFrontier(NodeType * node);
+
+  /** \brief Increments activation count; adds to frontier if threshold is met. */
+  void
+  incActivationCount(NodeType * node, std::size_t threshold);
+
+  /** \brief Decrements activation count; removes from frontier if threshold is no longer met. */
+  void
+  decActivationCount(NodeType * node, std::size_t threshold);
+
+  /** \brief Removes any state associated with the given node */
+  void
+  removeNode(NodeType * node);
+
+  NodeType *
+  peek();
+
+private:
+  using FrontierList = std::list<NodeType *>;
+
+  struct State
+  {
+    TraversalNodestate state = TraversalNodestate::ahead;
+    std::size_t activationCount = 0;
+    typename FrontierList::iterator pos = {};
+  };
+
+  std::unordered_map<NodeType *, State> states_;
+  FrontierList frontier_;
+};
+
+template<typename Traverser, typename NodeType>
 class TraverserIterator
 {
 public:
-  typedef std::input_iterator_tag iterator_category;
-  typedef Node * value_type;
-  typedef ssize_t difference_type;
-  typedef value_type * pointer;
-  typedef value_type & reference;
+  using iterator_category = std::input_iterator_tag;
+  using value_type = NodeType *;
+  using difference_type = ssize_t;
+  using pointer = value_type *;
+  using reference = value_type &;
 
-  constexpr explicit TraverserIterator(T * traverser = nullptr, Node * node = nullptr) noexcept
+  constexpr TraverserIterator(Traverser * traverser, NodeType * node) noexcept
```

Comment:
Both parameters could be references, no?

Reply:
done